features/storage/kvstore/conf/kv_config.cpp

```diff
@@ -0,0 +1,706 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "kv_config.h"
+#include "KVStore.h"
+#include "kv_map.h"
+#include "BlockDevice.h"
+#include "FileSystem.h"
+#include "FileSystemStore.h"
+#include "SlicingBlockDevice.h"
+#include "FATFileSystem.h"
+#include "LittleFileSystem.h"
+#include "TDBStore.h"
+#include "mbed_error.h"
+#include "FlashIAP.h"
+#include "FlashSimBlockDevice.h"
+#include "mbed_trace.h"
+#define TRACE_GROUP "KVCFG"
+
+#if COMPONENT_FLASHIAP
+#include "FlashIAPBlockDevice.h"
+#endif
+
+#if COMPONENT_QSPIF
+#include "QSPIFBlockDevice.h"
+#endif
+
+#if COMPONENT_SPIF
+#include "SPIFBlockDevice.h"
+#endif
+
+#if COMPONENT_DATAFLASH
+#include "DataFlashBlockDevice.h"
+#endif
+
+#if COMPONENT_SD
+#include "SDBlockDevice.h"
+#endif
+
+
+using namespace mbed;
+
+typedef struct {
+    KVStore *kvstroe_main_instance;
+    KVStore *internal_store;
+    KVStore *external_store;
+    BlockDevice *internal_bd;
+    BlockDevice *external_bd;
+    FileSystem *external_fs;
+} kvstore_config_t;
+
+static SingletonPtr<PlatformMutex> mutex;
+static bool is_kv_config_initialize = false;
+static kvstore_config_t kvstore_config;
+
+#define INTERNAL_BLOCKDEVICE_NAME FLASHIAP
+
+#define STR_EXPAND(tok) #tok
+#define STR(tok) STR_EXPAND(tok)
+
+#define _GET_FILESYSTEM_concat(dev, ...) _get_filesystem_##dev(__VA_ARGS__)
+#define GET_FILESYSTEM(dev, ...) _GET_FILESYSTEM_concat(dev, __VA_ARGS__)
+
+#define _GET_BLOCKDEVICE_concat(dev, ...) _get_blockdevice_##dev(__VA_ARGS__)
+#define GET_BLOCKDEVICE(dev, ...) _GET_BLOCKDEVICE_concat(dev, __VA_ARGS__)
+
+static inline uint32_t align_up(uint64_t val, uint64_t size)
+{
+    return (((val - 1) / size) + 1) * size;
+}
+
+static inline uint32_t align_down(uint64_t val, uint64_t size)
+{
+    return (((val) / size)) * size;
+}
+
+int _get_addresses(BlockDevice *bd, bd_addr_t start_address, bd_size_t size, bd_addr_t *out_start_addr,
+                   bd_addr_t *out_end_addr )
+{
+    bd_addr_t aligned_end_address;
+    bd_addr_t end_address;
+    bd_addr_t aligned_start_address;
+
+    aligned_start_address = align_down(start_address, bd->get_erase_size(start_address));
+    if (aligned_start_address != start_address) {
+        tr_error("KV Config: Start address is not aligned. Better use %02llx", aligned_start_address);
+        return -1;
+    }
+
+    if (size == 0) {
+        (*out_start_addr) = aligned_start_address;
+        (*out_end_addr) = bd->size();
+        return 0;
+    }
+
+    if (size != 0) {
+        end_address = start_address + size;
+        aligned_end_address = align_up(end_address, bd->get_erase_size(end_address));
+        if (aligned_end_address != end_address) {
+            tr_error("KV Config: End address is not aligned. Consider changing the size parameter.");
+            return -1;
+        }
+    }
+
+    if (aligned_end_address > bd->size()) {
```

Comment:
I think it would be clearer to move this into "if (size !=0)" and then I think you also don't need the "if (size ==0)" condition